Support antrea as network policy provider in kube-up

anfernee opened the pull request:
* Linux Support

Use the following environment to enable antrea on linux for kube-up:
```
export NETWORK_POLICY_PROVIDER=antrea
```

Currently, antrea is not supported on cos yet, so use ubuntu instead.
For example, the following environments:
```
export KUBE_NODE_OS_DISTRIBUTION=ubuntu
export KUBE_GCE_NODE_IMAGE=ubuntu-2004-focal-v20200423
export KUBE_GCE_NODE_PROJECT=ubuntu-os-cloud
```

* Windows support

Use the following environment to enable antrea on windows for kube-up:
```
export WINDOWS_NETWORK_POLICY_PROVIDER=antrea
```

This commit also allows to specify which windows image to use in the
cluster by
```
export WINDOWS_NODE_IMAGE_PROJECT=<project_name>
export WINDOWS_NODE_IMAGE=<image_name>
```

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
The PR allows to create antrea-enabled windows nodes with `kube-up.sh`. It only allows to create windows nodes now, because there is some problem running antrea on COS based images. 

```release-note
Support antrea as network policy provider in kube-up
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
n/a 

```docs
```

anfernee committed:
Support antrea as network policy in kube-up

* Linux Support

Use the following environment to enable antrea on linux for kube-up:
```
export NETWORK_POLICY_PROVIDER=antrea
```

Currently, antrea is not supported on cos yet, so use ubuntu instead.
For example, the following environments:
```
export KUBE_NODE_OS_DISTRIBUTION=ubuntu
export KUBE_GCE_NODE_IMAGE=ubuntu-2004-focal-v20200423
export KUBE_GCE_NODE_PROJECT=ubuntu-os-cloud
```

* Windows support

Use the following environment to enable antrea on windows for kube-up:
```
export WINDOWS_NETWORK_POLICY_PROVIDER=antrea
```

This commit also allows to specify which windows image to use in the
cluster by
```
export WINDOWS_NODE_IMAGE_PROJECT=<project_name>
export WINDOWS_NODE_IMAGE=<image_name>
```

anfernee commented:
cc @antoninbas @lzhecheng

anfernee commented:
/retest

dims commented:
@anfernee would it be possible to first get a linux / antrea change in? (before we try to do windows?). I'd expect a PR that introduces antrea on linux with network policy switched on, followed by a periodic CI job that exercises these changes and runs the network policy tests (and the results show up in test-grid).

Then we figure out how to do the same with windows. WDYT?

anfernee commented:
I tried linux version as well, but it failed to start on COS images. I don't know why yet (created https://github.com/vmware-tanzu/antrea/issues/2021). The change will help us testing antrea-windows compatibility, which could be used to build a verification pipeline for antrea. It would be very helpful for that task if this can be merged separately.

dims commented:
@anfernee you don't need to use COS images, you can use kube-up with ubuntu images too! 

example : https://github.com/kubernetes/test-infra/blob/7affdf2921a038c6d78e4a1f2f018f53c959893c/config/jobs/kubernetes/sig-cloud-provider/gcp/gcp-gce.yaml#L307-L310

dims commented:
@anfernee `/cluster` is a directory we are trying to deprecate/remove for a while, while we allow small/targeted things to merge, we do not encourage a large change like what is being proposed here ( see https://github.com/kubernetes/kubernetes/issues/78995 for context ). 

So let's see if there is a limited set of changes that will help run network policy tests on linux first.

cc @spiffxp @BenTheElder

cheftako commented on cluster/addons/antrea/linux/daemonset.yaml:
```diff
@@ -0,0 +1,208 @@
+apiVersion: apps/v1
+kind: DaemonSet
```
/cc @jkh52 
There are concerns with the total number of DaemonSets we are running.

cheftako replied:
Also the commit references windows but I notice this says linux in the path.

anfernee replied:
I repurposed this to support linux as well. @jkh52 what is the concern about the number of DaemonSets?

jkh52 replied:
In my experience with DaemonSets, when there are enough nodes (example: hundreds or more) and HA clusters (3 masters) the proxy-server could have runaway memory ballooning (at times like rolling master restarts, when we expect many agent grpc connections). I found the root cause to be: the agent authentication path in proxy-server was getting throttled by client-go and accumulating resources.

I found stability improvements by tuning:
- proxy-agent syncinterval
- proxy-server k8s client throttling (https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/192)

jeremyje replied:
Are there node startup scripts that this could be placed in? Or is this solving a chicken/egg problem where Antrea network fabric is not known to be selected before the node boots?

cheftako commented on cluster/gce/config-default.sh:
```diff
@@ -566,3 +573,18 @@ export WINDOWS_ENABLE_DSR="${WINDOWS_ENABLE_DSR:-false}"
 export TLS_CIPHER_SUITES=""
+
+# Optional: URL to download antrea-cni.exe for Windows node
+export WINDOWS_ANTREA_CNI_BINARY_URL="${WINDOWS_ANTREA_CNI_BINARY_URL:-https://github.com/vmware-tanzu/antrea/releases/download/v0.13.1/antrea-cni-windows-x86_64.exe}"
```
Do we want CNI and Agent at the same version? If so it may make sense to use a common version or URL base.

anfernee replied:
Most likely, but I am not assuming that. It could be different `.gke-x` versions in the future, so I would vote keep the flexibility there.

cheftako commented on cluster/gce/util.sh:
```diff
@@ -83,17 +83,20 @@ function set-linux-node-image() {
 function set-windows-node-image() {
-  WINDOWS_NODE_IMAGE_PROJECT="windows-cloud"
+  # If WINDOWS_NODE_IMAGE exits, use it.
+  if [[ "${WINDOWS_NODE_IMAGE}" != "" ]]; then
```
We use ```set -o nounset``` so I believe this needs to be 
```if [[ "${WINDOWS_NODE_IMAGE:-}" != "" ]]; then```

antoninbas commented on cluster/gce/windows/k8s-node-setup.psm1:
```diff
@@ -436,6 +442,121 @@ function DownloadAndInstall-CSIProxyBinaries {
+  # Download ovs installer
+  $tmp_dir = 'C:\k8s_tmp'
+  New-Item -Force -ItemType 'directory' $tmp_dir | Out-Null
+  $filename = 'Install-OVS.ps1'
```
In the latest version of the script, we added a new command-line option, `ImportCertificate` (https://github.com/vmware-tanzu/antrea/blob/002379983cbf685ffb1623cfb953d4443b18ef14/hack/windows/Install-OVS.ps1#L21). You can set it to `False` when you bring your own signed binaries.

mmottaghi replied:
* Sure, that could be a good solution, but even then, my other comment (line 483) still applies: input arguments must change.
* The other issue we have is that this file is not the right place to become so specific about the network policy component.
Here we have to only decide if we do or don't want to install a nwk-policy component, and if yes, we have to delegate the control to a more component-specific file (which would be Antrea installer in this case) to take care of the installation for us. If we directly install Antrea here, over time as more and more newer network components emerge this file will get polluted and messy with all the new installation calls.

anfernee replied:
I'll keep it as is. @mmottaghi can you add a patch to support officially signed ovs?

mmottaghi replied:
will do

jsturtevant commented:
fyi @jayunit100

jsturtevant commented:
/cc @jayunit100

anfernee commented:
Because of antrea's recent ga release, I am working on integrate it with ga. mark it as draft for now.

anfernee commented:
Hi folks,

I've modified the change to include 1.0.0 antrea release, and also make it work on Linux (Ubuntu). The official windows node still doesn't support `noEncap` mode, so I am using my private build with backport changes. It is not reliable yet. @antoninbas and I will be checking it out. 

This is the environments I am using:
```
export NETWORK_POLICY_PROVIDER=antrea
export WINDOWS_NETWORK_POLICY_PROVIDER=antrea

export NUM_NODES=2  # number of Linux nodes
export NUM_WINDOWS_NODES=1
export KUBE_GCE_ENABLE_IP_ALIASES=true
export KUBERNETES_NODE_PLATFORM=windows
export LOGGING_STACKDRIVER_RESOURCE_TYPES=new
export KUBE_UP_AUTOMATIC_CLEANUP=true
export KUBE_CONTAINER_RUNTIME=docker

# windows
export WINDOWS_NODE_OS_DISTRIBUTION=win2019
export WINDOWS_ANTREA_CNI_BINARY_URL='https://storage.googleapis.com/ygui-gke-dev/antrea-cni.exe'
export WINDOWS_ANTREA_AGENT_BINARY_URL='https://storage.googleapis.com/ygui-gke-dev/antrea-agent.exe'

# Linux images
export KUBE_GCE_MASTER_IMAGE=ubuntu-2004-focal-v20200423
export KUBE_GCE_MASTER_PROJECT=ubuntu-os-cloud
export KUBE_NODE_OS_DISTRIBUTION=ubuntu
export KUBE_GCE_NODE_IMAGE=ubuntu-2004-focal-v20200423
export KUBE_GCE_NODE_PROJECT=ubuntu-os-cloud
```

dims commented:
@anfernee please break this into two. Let's see how just the linux one looks first, then we can deal with the windows one later.

thanks!

anfernee commented:
Can we merge them together. They are equally going to be useful for @antoninbas to build more antrea pipelines.

dims commented:
@anfernee the `cluster/` directory is something we are trying to get rid of, so we try to keep the changes here minimal. So no, we will need to see what the linux one will look like first to even consider anything more.

antoninbas commented:
@dims if nothing new should be added to the `cluster/` directory, does splitting this PR between Linux support & Windows support achieve anything? What would be the rationale for accepting a Linux PR first?

dims commented:
@antoninbas i did not say "nothing new", i said "minimal" :) if the linux PR turns out to be small, we could do at least that.

anfernee commented:
cc @jeremyje

cheftako commented on cluster/gce/config-default.sh:
```diff
@@ -198,6 +198,12 @@ export MASTER_NODE_LABELS="${KUBE_MASTER_NODE_LABELS:-}"
+NETWORK_POLICY_PROVIDER="${NETWORK_POLICY_PROVIDER:-none}" # calico
+
+# Windows Network Policy plugin specific settings.
+export c="${WINDOWS_NETWORK_POLICY_PROVIDER:-none}"  # antrea
```
c? typo? I don't see this used later and it seems like a bad variable name.

anfernee replied:
Good catch! Fixed it.

BenTheElder commented:
> So let's see if there is a limited set of changes that will help run network policy tests on linux first.

Don't we already run network policy tests with kube-up on linux with calico? cc @aojea

aojea commented:
> > So let's see if there is a limited set of changes that will help run network policy tests on linux first.
> 
> Don't we already run network policy tests with kube-up on linux with calico? cc @aojea

yes we do, https://testgrid.k8s.io/sig-network-gce#network-policies,%20google-gce , it is Calico because the code was there since the beginning and we needed a way to test the new testing implemented in the e2e framework for network policies

I thought this was only for windows, I didn't know the goal was to set this for linux too ... 

I think we shouldn't build a CNI testing matrix in k/k ... there are more than 20 CNI plugins in the web https://kubernetes.io/docs/concepts/cluster-administration/networking/ , what happen if a new plugin wants to add its job too?

aojea commented on cluster/addons/antrea/common.yaml:
```diff
@@ -0,0 +1,3260 @@
+    - jsonPath: .metadata.creationTimestamp
+      name: Age
+      type: date
+    name: v1alpha1
```
this is v1alpah1, that doesnt sound like the kind of stability we need to gate on ...

aojea commented:
also, this has 26 fields in v1alpha1, is this stable enough to gate code in k/k?

anfernee commented:
[antrea](https://github.com/vmware-tanzu/antrea/releases/tag/v1.0.0) recently reached 1.0.0 milestone. @antoninbas Is there a plan to promote the api to `v1`? Regardless the tag, I believe the API should be pretty stable.

anfernee commented:
> > > So let's see if there is a limited set of changes that will help run network policy tests on linux first.
> > 
> > 
> > Don't we already run network policy tests with kube-up on linux with calico? cc @aojea
> 
> yes we do, https://testgrid.k8s.io/sig-network-gce#network-policies,%20google-gce , it is Calico because the code was there since the beginning and we needed a way to test the new testing implemented in the e2e framework for network policies
> 
> I thought this was only for windows, I didn't know the goal was to set this for linux too ...
> 
> I think we shouldn't build a CNI testing matrix in k/k ... there are more than 20 CNI plugins in the web https://kubernetes.io/docs/concepts/cluster-administration/networking/ , what happen if a new plugin wants to add its job too?

Original purpose is to support windows. There was a comment to enable that on Linux nodes as well, because Linux nodes are required, and it's strange to leave it behind. I think it makes sense, so added both.

antoninbas commented:
> [antrea](https://github.com/vmware-tanzu/antrea/releases/tag/v1.0.0) recently reached 1.0.0 milestone. @antoninbas Is there a plan to promote the api to `v1`? Regardless the tag, I believe the API should be pretty stable.

We do intend to promote the APIs to Beta and eventually GA, but we make a decision for each CRD separately for each new Antrea release. The Antrea policy API is still going through changes, even though we consider the implementation itself stable. Note that these Antrea-specific APIs are of course not exercised by any upstream test…

anfernee committed:
Remove v1alpha1 CRDs from the antrea

These CRDs are not required for basic functionality, so remove them.

dims commented:
/hold

https://github.com/kubernetes/kubernetes/pull/100736#issuecomment-828411329

aojea commented:
Let me suggest a different direction, why don't you use the new repo that the network policy folks are going to create?
https://github.com/kubernetes/community/issues/5747#issuecomment-828707975

You can use prow from there and we don't keep growing the tech debt in k/k , 
> I think we shouldn't build a CNI testing matrix in k/k ... there are more than 20 CNI plugins in the web https://kubernetes.io/docs/concepts/cluster-administration/networking/ , what happen if a new plugin wants to add its job too?

Also, we avoid this  ^^^,  Kubernetes network policy is the same as Ingress, it only defines the API not the implementation

anfernee commented:
> also, this has 26 fields in v1alpha1, is this stable enough to gate code in k/k?

I've removed the v1alpha1 APIs and changed the config to disable other features that relies on the alpha apis.

antoninbas commented on cluster/gce/windows/README-GCE-Windows-kube-up.md:
````diff
@@ -74,6 +74,40 @@ export KUBERNETES_NODE_PLATFORM=windows
 export LOGGING_STACKDRIVER_RESOURCE_TYPES=new
 ```
 
+**Experimental: Winodws**
````
s/Winodws/Windows

dims commented:
repeating again, please feel to experiment in this PR. but we will not merge this as mentioned several times earlier in this PR.

anfernee commented:
Hey @dims this is already the smallest change to support antrea. What do you suggest to make this mergeable?

jeremyje commented on cluster/addons/antrea/linux/daemonset.yaml:
```diff
@@ -0,0 +1,262 @@
+      hostNetwork: true
+      containers:
+        - name: node-init
+          image: gcr.io/google-containers/startup-script:v1
```
Use v2? v1 was built in 2016.

jeremyje commented on cluster/addons/antrea/linux/daemonset.yaml:
```diff
@@ -0,0 +1,262 @@
+          valueFrom:
+            fieldRef:
+              fieldPath: spec.nodeName
+        image: projects.registry.vmware.com/antrea/antrea-ubuntu:v1.0.0
```
Should these images be hosted from a canonical Kubernetes repository?

jeremyje commented on cluster/gce/config-default.sh:
```diff
@@ -167,7 +167,7 @@ export ENABLE_DOCKER_REGISTRY_CACHE=true
 
 # Optional: Deploy a L7 loadbalancer controller to fulfill Ingress requests:
 #   glbc           - CE L7 Load Balancer Controller
-export ENABLE_L7_LOADBALANCING="${KUBE_ENABLE_L7_LOADBALANCING:-glbc}"
```
Bad merge? Why is this getting deleted? Or is the comment above meant to be preserved?

jeremyje commented on cluster/gce/util.sh:
```diff
@@ -83,17 +83,20 @@ function set-linux-node-image() {
 # Requires:
 #   WINDOWS_NODE_OS_DISTRIBUTION
 # Sets:
-#   WINDOWS_NODE_IMAGE_PROJECT
```
Why is the set-windows-node-image function being changed? I wouldn't expect that to be updated.

jeremyje commented on cluster/gce/gci/node.yaml:
```diff
@@ -117,6 +117,25 @@ write_files:
       [Install]
       WantedBy=multi-user.target
 
+  - path: /lib/systemd/system/google-network-daemon.service
```
Will google-network-daemon be installed as a systemd service on all COS/GCI images going forward even when Antrea isn't installed or is there a flag somewhere that ensures this is only active when Antrea is enabled.

jeremyje commented on cluster/gce/windows/k8s-node-setup.psm1:
```diff
@@ -1146,6 +1346,18 @@ function Configure-Kubelet {
+  AntreaPolicy: false
+  Traceflow: false
+clientConnection:
+  kubeconfig: \etc\kubernetes\antrea.kubeconfig
```
prefix with `C:` or use `/` delimiters.

mmottaghi replied:
@jeremyje  `\etc\kubernetes\antrea.kubeconfig` works. 
This is the way the Antrea team uses it.

jeremyje commented on cluster/gce/windows/k8s-node-setup.psm1:
```diff
@@ -966,6 +1140,9 @@ function Prepare-CniNetworking {
     Configure_Dockerd_CniNetworking
   }
+  if (${env:WINDOWS_NETWORK_POLICY_PROVIDER} -eq "antrea") {
+    Configure_Antrea_CniNetworking
```
Is this tested with containerd as well?

liggitt commented on cluster/addons/antrea/common.yaml:
```diff
@@ -0,0 +1,1616 @@
+    rbac.authorization.k8s.io/aggregate-to-edit: "true"
+  name: aggregate-antrea-clustergroups-edit
+rules:
+- apiGroups:
+  - core.antrea.tanzu.vmware.com
+  resources:
+  - clustergroups
+  verbs:
+  - get
+  - list
+  - watch
+  - create
+  - update
+  - patch
+  - delete
+- apiGroups:
+  - crd.antrea.io
+  resources:
+  - clustergroups
+  verbs:
+  - get
+  - list
+  - watch
+  - create
+  - update
+  - patch
+  - delete
```
what is this addon doing with these permissions (and all the CRUD permissions below). this seems like an very wide set of permissions

liggitt reviewed (commented):
> I tried linux version as well, but it failed to start on COS images. I don't know why yet (created vmware-tanzu/antrea#2021). The change will help us testing antrea-windows compatibility, which could be used to build a verification pipeline for antrea. It would be very helpful for that task if this can be merged separately.

In my opinion, new addons and callers should not be added to this directory. Every new thing added is one more obstacle to cleaning up it and removing it.

anfernee commented:
There is ongoing discussion where this change should land, so mark it as draft.

anfernee commented:
@all It looks like this `cluster` folder as well as all the cloud provider specific change is going to phase out from this repo. All the new features are forbidden. As a result, this change won't be suitable here. In the long run, https://github.com/kubernetes/cloud-provider-gcp is a better place for this change. That repo is still a little bit behind, but it will catch up. We will keep working with antrea and kubernetes community to make them work better together. Thanks a lot for the reviewers who have put a lot of time on this change.
